### ISSUE TITLE
为什么执行了python3 main.py --do_train --do_eval后result.txt文件是空的？

### ISSUE THREAD

**Issue by yangxia605** [monologg/R-BERT#12]
你好：
请教一下，为什么执行了python3 main.py --do_train --do_eval之后，result.txt 文件是空的，请问这是什么原因呢？谢谢

**Comment by wenston123** [comment]
你好，您是如何执行--do eval，do--train之后没有生成模型文件啊

**Comment by Pvvv** [comment]
> 你好：
> 请教一下，为什么执行了python3 main.py --do_train --do_eval之后，result.txt 文件是空的，请问这是什么原因呢？谢谢

请问您的问题现在解决了嘛

**Comment by hialoha** [comment]
I also have the same problem,result.txt is empty,and I haven't deal with it.

**Comment by Claymore715** [comment]
我也有这个问题

**Comment by Pvvv** [comment]
得下载一个 perl&nbsp; &nbsp;才可以


-----------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "monologg/R-BERT"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年5月20日(星期四) 晚上9:27
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [monologg/R-BERT] 为什么执行了python3 main.py --do_train --do_eval后result.txt文件是空的？ (#12)





 
我也有这个问题

**Comment by wjx-git** [comment]
除了安装Perl外，还需要将 eval/answer_keys.txt 中的标签改为你自己验证集数据的标签。还需要更改 semeval2010_task8_scorer-v1.2.pl 中的部分内容，挺麻烦的，还不如用 sklearn 计算。

**Comment by lixianglong1205** [comment]
# 使用sklearn计算F1-score
# 欢迎指正
import os
from sklearn.metrics import f1_score, precision_score, accuracy_score, recall_score

EVAL_DIR = "eval"


# def official_f1():
#     # Run the perl script
#     try:
#         cmd = "perl {0}/semeval2010_task8_scorer-v1.2.pl {0}/proposed_answers.txt {0}/answer_keys.txt > {0}/result.txt".format(
#             EVAL_DIR
#         )
#         os.system(cmd)
#     except:
#         raise Exception("perl is not installed or proposed_answers.txt is missing")
#
#     with open(os.path.join(EVAL_DIR, "result.txt"), "r", encoding="utf-8") as f:
#         macro_result = list(f)[-1]
#         macro_result = macro_result.split(":")[1].replace(">>>", "").strip()
#         macro_result = macro_result.split("=")[1].strip().replace("%", "")
#         macro_result = float(macro_result) / 100
#
#     return macro_result


def read_lael(path="./data/label.txt"):
    with open(path, mode="r", encoding="utf-8") as file:
        label_list = [i.strip() for i in file]
    results = {}
    for idx, label in enumerate(label_list):
        results[label] = idx
    return results


def txt_result_analysis(path):
    with open(path, mode="r", encoding="utf-8") as answer_file:
        source_label = [i.rstrip() for i in answer_file if i.strip()]
    label_dict = {}
    for one in source_label:
        i, j = one.split("\t")
        label_dict[i] = j

    return label_dict


def official_f1(average="macro"):
    """使用sklearn包计算f1-score"""
    label_dict = read_lael()
    source_label_dict = txt_result_analysis("./eval/answer_keys.txt")
    proposed_label_dict = txt_result_analysis("./eval/proposed_answers.txt")
    source_label, proposed_label = [], []
    for key in source_label_dict:
        value1 = source_label_dict[key]
        value2 = proposed_label_dict[key]
        source_label.append(value1)
        proposed_label.append(value2)
        # source_label.append(label_dict[value1])
        # proposed_label.append(label_dict[value2])
    # micro: 计算所有的TP TN FP FN，然后计算F1-score
    # f1_score_ = f1_score(source_label, proposed_label, average="micro")
    # # macro: 计算每一个类别的F1-score，然后对所有的F1-scre求平均值
    # f1_score_ = f1_score(source_label, proposed_label, average="macro")
    f1_score_ = f1_score(source_label, proposed_label, average=average)
    return f1_score_


if __name__ == "__main__":
    # official_f1()
    print("macro-averaged F1 = {}%".format(official_f1() * 100))

**Comment by lixianglong1205** [comment]
# 使用sklearn计算F1-score
# 欢迎指正
`import os
from sklearn.metrics import f1_score, precision_score, accuracy_score, recall_score

EVAL_DIR = "eval"


# def official_f1():
#     # Run the perl script
#     try:
#         cmd = "perl {0}/semeval2010_task8_scorer-v1.2.pl {0}/proposed_answers.txt {0}/answer_keys.txt > {0}/result.txt".format(
#             EVAL_DIR
#         )
#         os.system(cmd)
#     except:
#         raise Exception("perl is not installed or proposed_answers.txt is missing")
#
#     with open(os.path.join(EVAL_DIR, "result.txt"), "r", encoding="utf-8") as f:
#         macro_result = list(f)[-1]
#         macro_result = macro_result.split(":")[1].replace(">>>", "").strip()
#         macro_result = macro_result.split("=")[1].strip().replace("%", "")
#         macro_result = float(macro_result) / 100
#
#     return macro_result


def read_lael(path="./data/label.txt"):
    with open(path, mode="r", encoding="utf-8") as file:
        label_list = [i.strip() for i in file]
    results = {}
    for idx, label in enumerate(label_list):
        results[label] = idx
    return results


def txt_result_analysis(path):
    with open(path, mode="r", encoding="utf-8") as answer_file:
        source_label = [i.rstrip() for i in answer_file if i.strip()]
    label_dict = {}
    for one in source_label:
        i, j = one.split("\t")
        label_dict[i] = j

    return label_dict


def official_f1(average="macro"):
    """使用sklearn包计算f1-score"""
    label_dict = read_lael()
    source_label_dict = txt_result_analysis("./eval/answer_keys.txt")
    proposed_label_dict = txt_result_analysis("./eval/proposed_answers.txt")
    source_label, proposed_label = [], []
    for key in source_label_dict:
        value1 = source_label_dict[key]
        value2 = proposed_label_dict[key]
        source_label.append(value1)
        proposed_label.append(value2)
        # source_label.append(label_dict[value1])
        # proposed_label.append(label_dict[value2])
    # micro: 计算所有的TP TN FP FN，然后计算F1-score
    # f1_score_ = f1_score(source_label, proposed_label, average="micro")
    # # macro: 计算每一个类别的F1-score，然后对所有的F1-scre求平均值
    # f1_score_ = f1_score(source_label, proposed_label, average="macro")
    f1_score_ = f1_score(source_label, proposed_label, average=average)
    return f1_score_


if __name__ == "__main__":
    # official_f1()
    print("macro-averaged F1 = {}%".format(official_f1() * 100))
`

**Comment by lixianglong1205** [comment]
`import os
from sklearn.metrics import f1_score, precision_score, accuracy_score, recall_score

EVAL_DIR = "eval"


def txt_result_analysis(path):
    with open(path, mode="r", encoding="utf-8") as answer_file:
        source_label = [i.rstrip() for i in answer_file if i.strip()]
    label_dict = {}
    for one in source_label:
        i, j = one.split("\t")
        label_dict[i] = j

    return label_dict


def official_f1(average="macro"):
    """使用sklearn包计算f1-score"""
    source_label_dict = txt_result_analysis("./eval/answer_keys.txt")
    proposed_label_dict = txt_result_analysis("./eval/proposed_answers.txt")
    source_label, proposed_label = [], []
    for key in source_label_dict:
        value1 = source_label_dict[key]
        value2 = proposed_label_dict[key]
        source_label.append(value1)
        proposed_label.append(value2)
        # source_label.append(label_dict[value1])
        # proposed_label.append(label_dict[value2])
    # micro: 计算所有的TP TN FP FN，然后计算F1-score
    # f1_score_ = f1_score(source_label, proposed_label, average="micro")
    # # macro: 计算每一个类别的F1-score，然后对所有的F1-scre求平均值
    # f1_score_ = f1_score(source_label, proposed_label, average="macro")
    f1_score_ = f1_score(source_label, proposed_label, average=average)
    return f1_score_


if __name__ == "__main__":
    # official_f1()
    print("macro-averaged F1 = {}%".format(official_f1() * 100))
`

**Comment by lixianglong1205** [comment]
使用sklearn计算f1-score
欢迎指正


`import os
from sklearn.metrics import f1_score, precision_score, accuracy_score, recall_score

EVAL_DIR = "eval"


def txt_result_analysis(path):
    with open(path, mode="r", encoding="utf-8") as answer_file:
        source_label = [i.rstrip() for i in answer_file if i.strip()]
    label_dict = {}
    for one in source_label:
        i, j = one.split("\t")
        label_dict[i] = j

    return label_dict


def official_f1(average="macro"):
    """使用sklearn包计算f1-score"""
    source_label_dict = txt_result_analysis("./eval/answer_keys.txt")
    proposed_label_dict = txt_result_analysis("./eval/proposed_answers.txt")
    source_label, proposed_label = [], []
    for key in source_label_dict:
        value1 = source_label_dict[key]
        value2 = proposed_label_dict[key]
        source_label.append(value1)
        proposed_label.append(value2)
    """
    micro: 计算所有的TP TN FP FN，然后计算F1-score
    f1_score_ = f1_score(source_label, proposed_label, average="micro")
    macro: 计算每一个类别的F1-score，然后对所有的F1-scre求平均值
    f1_score_ = f1_score(source_label, proposed_label, average="macro")
    """
    f1_score_ = f1_score(source_label, proposed_label, average=average)
    return f1_score_


if __name__ == "__main__":
    print("macro-averaged F1 = {}%".format(official_f1() * 100))
`

**Comment by lixianglong1205** [comment]
```python
import os
from sklearn.metrics import f1_score, precision_score, accuracy_score, recall_score

EVAL_DIR = "eval"


def txt_result_analysis(path):
    with open(path, mode="r", encoding="utf-8") as answer_file:
        source_label = [i.rstrip() for i in answer_file if i.strip()]
    label_dict = {}
    for one in source_label:
        i, j = one.split("\t")
        label_dict[i] = j

    return label_dict


def official_f1(average="macro"):
    """使用sklearn包计算f1-score"""
    source_label_dict = txt_result_analysis("./eval/answer_keys.txt")
    proposed_label_dict = txt_result_analysis("./eval/proposed_answers.txt")
    source_label, proposed_label = [], []
    for key in source_label_dict:
        value1 = source_label_dict[key]
        value2 = proposed_label_dict[key]
        source_label.append(value1)
        proposed_label.append(value2)
    """
    micro: 计算所有的TP TN FP FN，然后计算F1-score
    f1_score_ = f1_score(source_label, proposed_label, average="micro")
    macro: 计算每一个类别的F1-score，然后对所有的F1-scre求平均值
    f1_score_ = f1_score(source_label, proposed_label, average="macro")
    """
    f1_score_ = f1_score(source_label, proposed_label, average=average)
    return f1_score_


if __name__ == "__main__":
    print("macro-averaged F1 = {}%".format(official_f1() * 100))

```

**Comment by lixianglong1205** [comment]
使用sklearn计算f1-score
欢迎指正
```pyhon
import os
from sklearn.metrics import f1_score, precision_score, accuracy_score, recall_score

EVAL_DIR = "eval"


def txt_result_analysis(path):
    with open(path, mode="r", encoding="utf-8") as answer_file:
        source_label = [i.rstrip() for i in answer_file if i.strip()]
    label_dict = {}
    for one in source_label:
        i, j = one.split("\t")
        label_dict[i] = j

    return label_dict


def official_f1(average="macro"):
    """使用sklearn包计算f1-score"""
    source_label_dict = txt_result_analysis("./eval/answer_keys.txt")
    proposed_label_dict = txt_result_analysis("./eval/proposed_answers.txt")
    source_label, proposed_label = [], []
    for key in source_label_dict:
        value1 = source_label_dict[key]
        value2 = proposed_label_dict[key]
        source_label.append(value1)
        proposed_label.append(value2)
    """
    micro: 计算所有的TP TN FP FN，然后计算F1-score
    f1_score_ = f1_score(source_label, proposed_label, average="micro")
    macro: 计算每一个类别的F1-score，然后对所有的F1-scre求平均值
    f1_score_ = f1_score(source_label, proposed_label, average="macro")
    """
    f1_score_ = f1_score(source_label, proposed_label, average=average)
    return f1_score_


if __name__ == "__main__":
    print("macro-averaged F1 = {}%".format(official_f1() * 100))


```
